File: cpp/src/Survey.cpp
#include <iostream> 
#include <fstream>
#include <nlohmann/json.hpp>
#include <sstream>
#include <vector>
#include <cmath>
#include "Survey.hpp"

using json = nlohmann::json;

const float pi = 3.14159265358979323846f;

Survey::Survey()
{
    readParameters();
    readGeometry();
}

void read_csv(std::string path, std::vector<float>& x,std::vector<float>& z)
{
    std::ifstream file(path);
    if (file.is_open()){
        std::string line;
        getline(file, line);
        while (getline(file, line)){
            std::stringstream ss(line);
            std::string index;
            std::string coordx;
            std::string coordz;
            getline(ss, index, ',');
            getline(ss, coordx, ',');
            getline(ss, coordz, ',');
            x.push_back(std::stof(coordx));
            z.push_back(std::stof(coordz));
        }
    }
}

void Survey::readParameters() 
{
    std::ifstream jsonFile("Parameters.json");
    json parameters = json::parse(jsonFile);

    unit = parameters["unit"].get<std::string>();
    approximation = parameters["approximation"].get<std::string>();
    migration = parameters["migration"].get<std::string>();
    ABC = parameters["ABC"].get<std::string>();
    dx = parameters["dx"].get<float>();
    dz = parameters["dz"].get<float>();
    dt = parameters["dt"].get<float>();
    L = parameters["L"].get<float>();
    D = parameters["D"].get<float>();
    T = parameters["T"].get<float>();
    N_abc = parameters["N_abc"].get<int>();
    fcut = parameters["fcut"].get<float>();
    sigma = parameters["sigma"].get<float>();
    dvel = parameters["dvel"].get<float>();
    ratio = parameters["ratio"].get<float>();
    shift = parameters["shift"].get<float>();
    window = parameters["window"].get<float>();
    v0 = parameters["v0"].get<float>();
    step = parameters["step"].get<int>();
    last_save = parameters["last_save"].get<int>();
    fwi = parameters["fwi"].get<bool>();
    niter = parameters["niter"].get<int>();
    freqs = parameters["freqs"].get<std::vector<float>>();
    vmin = parameters["vmin"].get<float>();
    vmax = parameters["vmax"].get<float>();
    epsmin = parameters["epsmin"].get<float>();
    epsmax = parameters["epsmax"].get<float>();
    deltamin = parameters["deltamin"].get<float>();
    deltamax = parameters["deltamax"].get<float>();
    thetamin = parameters["thetamin"].get<float>();
    thetamax = parameters["thetamax"].get<float>();
    multiparameter = parameters["multiparameter"].get<bool>();
    reciprocity = parameters["reciprocity"].get<bool>();
    mirror = parameters["mirror"].get<bool>();
    layer2 = parameters["layer2"].get<bool>();
    layer3 = parameters["layer3"].get<bool>();
    gradientmodel = parameters["gradientmodel"].get<bool>();
    diffractor = parameters["diffractor"].get<bool>();
    modelfromvp = parameters["modelfromvp"].get<bool>();
    waterlayer = parameters["waterlayer"].get<bool>();
    idx_water = parameters["idx_water"].get<int>();
    snap = parameters["snap"].get<bool>();
    rec_file = parameters["rec_file"].get<std::string>();
    src_file = parameters["src_file"].get<std::string>();
    vpFile = parameters["vpFile"].get<std::string>();
    epsilonFile = parameters["epsilonFile"].get<std::string>();
    deltaFile = parameters["deltaFile"].get<std::string>();
    thetaFile = parameters["thetaFile"].get<std::string>();
    snapshotFolder = parameters["snapshotFolder"].get<std::string>();
    seismogramFolder = parameters["seismogramFolder"].get<std::string>();
    checkpointFolder = parameters["checkpointFolder"].get<std::string>();
    migratedimageFolder = parameters["migratedimageFolder"].get<std::string>();
    modelFolder = parameters["modelFolder"].get<std::string>();
    estimatedmodelsFolder = parameters["estimatedmodelsFolder"].get<std::string>();
    gradientsFolder = parameters["gradientsFolder"].get<std::string>();

    tlag = 2.0f * std::sqrt(pi)/fcut;
    itlag = std::round(tlag/dt);
    nt_data = std::round(T/dt) + 1;
    nx = std::round(L/dx) + 1;
    nz = std::round(D/dz) + 1;
    nt = itlag + nt_data;
    nx_abc = nx + 2*N_abc;
    nz_abc = nz + 2*N_abc;

    x.resize(nx);
    z.resize(nz);
    t.resize(nt);

    for (int i = 0; i < nx; i++) {
        x[i] = i * dx;
    }

    for (int i = 0; i < nz; i++) {
        z[i] = i * dz;
    }

    for (int i = 0; i < nt; i++) {
        t[i] = i * dt;
    }

}

void Survey::readGeometry()
{
    read_csv(rec_file, rec_x, rec_z);
    read_csv(src_file, shot_x, shot_z);

    if (reciprocity) {
        std::swap(shot_x, rec_x);
        std::swap(shot_z, rec_z);
    }

    Nrec = (rec_x.size());
    Nshot = (shot_x.size());

    rx.resize(Nrec);
    rz.resize(Nrec);

    sx.resize(Nshot);
    sz.resize(Nshot);

    for (int irec = 0; irec < Nrec; ++irec)
    {
        rx[irec] = std::round(rec_x[irec] / dx) + N_abc;
        rz[irec] = std::round(rec_z[irec] / dz) + N_abc;
    }

    for (int ishot = 0; ishot < Nshot; ++ishot)
    {
        sx[ishot] = std::round(shot_x[ishot] / dx) + N_abc;
        sz[ishot] = std::round(shot_z[ishot] / dz) + N_abc;
    }

    if (mirror)
    {
        for (int irec = 0; irec < Nrec; ++irec)
        {
            rz[irec] = std::round(rec_z[irec] / dz) + N_abc - idx_water;
        }

        for (int ishot = 0; ishot < Nshot; ++ishot)
        {
            sz[ishot] = std::round(shot_z[ishot] / dz) + N_abc - idx_water;
        }
    }    
    
}

File: src/survey.py
import pandas as pd
import json
import numpy as np

class parameters: 
    def __init__(self, parameters_path):
        self.parameters_path = parameters_path
        self.readParameters()
        self.readAcquisitionGeometry()

    def readParameters(self):
        with open(self.parameters_path) as f:
            self.parameters = json.load(f)

        # Approximation type
        self.approximation = self.parameters["approximation"]

        
        # Discretization self.parameters
        self.dx   = self.parameters["dx"]
        self.dz   = self.parameters["dz"]
        self.dt   = self.parameters["dt"]
        
        # Model size
        self.L    = self.parameters["L"]
        self.D    = self.parameters["D"]
        self.T    = self.parameters["T"]

        # Number of point for absorbing boundary condition
        self.N_abc = self.parameters["N_abc"]

        # Max frequency
        self.fcut = self.parameters["fcut"]

        # Source delay
        self.tlag = 2.0*np.sqrt(np.pi)/self.fcut

        # Number of points in each direction
        self.nx = np.round(self.L/self.dx).astype(np.int32)+1
        self.nz = np.round(self.D/self.dz).astype(np.int32)+1
        self.itlag = np.round(self.tlag / self.dt).astype(np.int32)
        self.nt_data = np.round(self.T / self.dt).astype(np.int32) + 1
        self.nt = self.itlag + self.nt_data

        self.nx_abc = self.nx + 2*self.N_abc
        self.nz_abc = self.nz + 2*self.N_abc

        # Define arrays for space and time
        self.x = np.arange(self.nx) * self.dx
        self.z = np.arange(self.nz) * self.dz
        self.t = np.arange(self.nt) * self.dt

        # Folders
        self.seismogramFolder = self.parameters["seismogramFolder"]
        self.imageFolder = self.parameters["imageFolder"]
        self.snapshotFolder = self.parameters["snapshotFolder"]
        self.modelFolder = self.parameters["modelFolder"]
        self.checkpointFolder = self.parameters["checkpointFolder"]
        self.estimatedmodelsFolder = self.parameters["estimatedmodelsFolder"]
        self.gradientsFolder = self.parameters["gradientsFolder"]

        # Source and receiver files
        self.rec_file = self.parameters["rec_file"]
        self.src_file = self.parameters["src_file"]

        # Velocity model file
        self.vpFile = self.parameters["vpFile"]
        self.thetaFile = self.parameters["thetaFile"]

        # Snapshot flag
        self.snap = self.parameters["snap"]
        self.step = self.parameters["step"]
        self.last_save = self.parameters["last_save"]

        # Anisotropy parameters files
        self.epsilonFile = self.parameters["epsilonFile"]  
        self.deltaFile   = self.parameters["deltaFile"]  

        # Synthetic models 
        self.layer2 =  self.parameters['layer2']
        self.layer3 =  self.parameters['layer3']
        self.gradientmodel =  self.parameters['gradientmodel']
        self.diffractor =  self.parameters['diffractor']
        self.modelfromvp =  self.parameters['modelfromvp']
        self.waterlayer = self.parameters['waterlayer']
        
        #migration parameters
        self.mirror = self.parameters['mirror']
        self.reciprocity = self.parameters['reciprocity']
        self.idx_water = self.parameters['idx_water']

    def readAcquisitionGeometry(self):        
        # Read receiver and source coordinates from CSV files
        receiverTable = pd.read_csv(self.rec_file)
        print(f"info: Imported: {self.rec_file}")     
        sourceTable = pd.read_csv(self.src_file)
        print(f"info: Imported: {self.src_file}")

        # Read receiver and source coordinates
        self.rec_x = receiverTable['coordx'].to_numpy()
        self.rec_z = receiverTable['coordz'].to_numpy()
        self.shot_x = sourceTable['coordx'].to_numpy()
        self.shot_z = sourceTable['coordz'].to_numpy()

        if self.reciprocity == True:
            self.shot_x, self.rec_x = self.rec_x.copy(), self.shot_x.copy()
            self.shot_z, self.rec_z = self.rec_z.copy(), self.shot_z.copy()

        self.rx = np.round(self.rec_x / self.dx).astype(np.int32) + self.N_abc 
        self.rz = np.round(self.rec_z / self.dz).astype(np.int32) + self.N_abc 
        self.sx = np.round(self.shot_x / self.dx).astype(np.int32) + self.N_abc
        self.sz = np.round(self.shot_z / self.dz).astype(np.int32) + self.N_abc 
            
        if self.mirror == True:
            self.rz = np.round(self.rec_z/self.dz).astype(np.int32) + self.N_abc - self.idx_water
            self.sz = np.round(self.shot_z/self.dz).astype(np.int32) + self.N_abc - self.idx_water

        self.Nrec = len(self.rec_x)
        self.Nshot = len(self.shot_x) 



    
  